journal: Add compact mode

daandemeyer opened the pull request:
The following improvements are gated behind the new compact mode:

- 32-bit offsets for Entry Array object payloads
- Use a Trie to store Entry object offsets
- Support for inlining specific fields (and their values) pairs into the Entry object (defaults to MESSAGE only)
- Support for not indexing specific fields
- Boot ID deduplication

The Trie commit contains a more detailed explanation of how it works. The changes should be backwards compatible. journalctl can both read new journals written by the updated systemd-journald and the current journals generated by a Fedora 35 system.

All credits go to @mokomull for the trie idea. The original experiment which I started from can be found here: https://github.com/mokomull/systemd/commit/1b1b3d06b55293526de425cb5696b35a70468e27

daandemeyer committed:
journal: Add compact mode

This adds a new flag in preparation for incompatible journal changes
which will be gated behind this flag. The max file size of journal
files in compact mode is limited to 4 GiB.

daandemeyer committed:
journal: Use 32-bit entry array offsets in compact mode

Before:

OBJECT TYPE      ENTRIES SIZE
Unused           0       0B
Data             3610336 595.7M
Field            5310    285.2K
Entry            3498326 1.2G
Data Hash Table  29	 103.1M
Field Hash Table 29      151.3K
Entry Array      605991  1011.6M
Tag              0	 0B
Total            7720021 2.9G

After:

OBJECT TYPE      ENTRIES SIZE
Unused           0	 0B
Data             3562667 591.0M
Field            3971    213.6K
Entry            3498566 1.2G
Data Hash Table  20	 71.1M
Field Hash Table 20	 104.3K
Entry Array	 582647  505.0M
Tag              0	 0B
Total            7647891 2.4G

daandemeyer committed:
journal: Enable compact mode

We also add an environment variable $SYSTEMD_JOURNAL_COMPACT that
can be used to disable compact mode if needed (similar to
$SYSTEMD_JOURNAL_KEYED_HASH).

daandemeyer committed:
journal: Store entry item offsets in trie in compact mode

In a journal file, data objects such as
"_BOOT_ID=100b7ef4c5124bc6839e28db763b4627" and "PRIORITY=6" are
referenced repeatedly; usually the only difference between two entries
is the "MESSAGE=" data object.  This results in a large amount of wasted
space, as each entry contains an array of 16-byte EntryItem pointers to
such data objects.

Instead, compress that array of items into a Trie, allowing us to only
record the difference between them:

node: pointer to _BOOT_ID=100b7ef4c5124bc6839e28db763b4627 data
  `- node: pointer to PRIORITY=6 data
      `- node: pointer to MESSAGE=foo
      `- node: pointer to MESSAGE=bar

The new Trie Node object stores an offset to the parent node, an offset
to the corresponding Data object, the next object in the hash table chain
and the hash itself. The hash we use is the XOR hash of all the Entry
items starting from the current node. In the best case scenario, this
allows us to find the correct node with just a single hash lookup.

If we assume that for most messages, all fields except MESSAGE have
already appeared in a previous message, we can compare the overhead
before and after this change per message:

(I'm not counting the EntryItem hash field as overhead here as we
can remove those regardless of whether we use a trie or not)

Before:
- A new Data object for the MESSAGE field
- A new Entry object with +- 20 64-bit offsets

After:
- A new Data object for the MESSAGE field
- A new Trie Node object for the MESSAGE field (6 * 64-bit)
- A new Entry object with a single 64-bit offset

So the Entry overhead reduces from 20 * 64-bit to 7 * 64-bit. If
we add opt-in field indexing in the future, we'll store the MESSAGE
field inline which will reduce the Entry overhead to a single 64-bit
offset for most messages.

Before:

OBJECT TYPE	 ENTRIES SIZE
Unused           0       0B
Data             3562667 591.0M
Field            3971    213.6K
Entry            3498566 1.2G
Data Hash Table  20      71.1M
Field Hash Table 20      104.3K
Entry Array      582647  505.0M
Tag              0       0B
Total            7647891 2.4G

After:

OBJECT TYPE      ENTRIES  SIZE
Unused           0        0B
Data             3521895  587.0M
Field            3140     169.4K
Entry            3499118  240.2M
Data Hash Table  14       49.7M
Field Hash Table 14       73.0K
Entry Array      577350   499.5M
Tag              0        0B
Trie Node        5767903  220.0M
Trie Hash Table  14       74.6M
Total            13369448 1.6G

daandemeyer committed:
journal: Introduce journal_file_entry_item_next() and journal_file_data_payload()

journal_file_entry_item_next() abstracts iterating over the entry items
of an entry object. It will allow us to introduce trie storage for entry
items and opt-in field indexing in later commits without having to make
large changes across the sd-journal codebase. Instead, most of the read
path changes will be localized to journal_file_entry_next().

journal_file_data_payload() retrieves the payload of a Data object,
optionally decompressing it and checking to see if matches a given
field. This function replaces all the decompression code in the sd-journal
codebase with a single function. The trie commit will also make use
of it and the opt-in field indexing commit will make use of the
maybe_decompress_payload() helper function.

This commit should not introduce any changes in sd-journal behavior.

daandemeyer committed:
journal: Introduce EntryItemEx

In later commits, we need to attach extra information to EntryItem.
To allow doing so, introduce a new struct EntryItemEx that we can
safely modify without modifying the journal format. We'll add more
fields to EntryItemEx in later commits.

medhefgo commented on src/libsystemd/sd-journal/journal-def.h:
```diff
@@ -153,30 +175,32 @@ enum {
+#  define HEADER_INCOMPATIBLE_SUPPORTED (HEADER_INCOMPATIBLE_COMPRESSED_ZSTD|HEADER_INCOMPATIBLE_KEYED_HASH|HEADER_INCOMPATIBLE_COMPACT)
 #else
-#  define HEADER_INCOMPATIBLE_SUPPORTED HEADER_INCOMPATIBLE_KEYED_HASH
+#  define HEADER_INCOMPATIBLE_SUPPORTED (HEADER_INCOMPATIBLE_KEYED_HASH|HEADER_INCOMPATIBLE_COMPACT)
```
All this ifdefery could be simplified:
```c
#define HEADER_INCOMPATIBLE_SUPPORTED                            \
        ((HAVE_XZ ? HEADER_INCOMPATIBLE_COMPRESSED_XZ : 0) |     \
         (HAVE_LZ4 ? HEADER_INCOMPATIBLE_COMPRESSED_LZ4 : 0) |   \
         (HAVE_ZSTD ? HEADER_INCOMPATIBLE_COMPRESSED_ZSTD : 0) | \
         HEADER_INCOMPATIBLE_KEYED_HASH |                        \
         HEADER_INCOMPATIBLE_COMPACT)
```

medhefgo commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -512,6 +520,8 @@ static bool warn_wrong_flags(const JournalFile *f, bool compatible) {
                                         strv[n++] = "zstd-compressed";
                                 if (flags & HEADER_INCOMPATIBLE_KEYED_HASH)
                                         strv[n++] = "keyed-hash";
+                                if (flags & HEADER_INCOMPATIBLE_COMPACT)
```
FLAGS_SET?

daandemeyer replied:
I think we only recommend using it as part of a larger expression. On it's own there's not really a lot of benefit to using it.

medhefgo commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -918,6 +947,26 @@ static int journal_file_check_object(JournalFile *f, uint64_t offset, Object *o)
 
                 break;
+
+        case OBJECT_TRIE_NODE:
```
The check for OBJECT_ENTRY_ARRAY further up doesn't take compact headers into account (same in journal-verify)

medhefgo commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1781,7 +2147,17 @@ uint64_t journal_file_entry_array_n_items(Object *o) {
-        return (sz - offsetof(Object, entry_array.items)) / sizeof(uint64_t);
+        return JOURNAL_HEADER_COMPACT(f->header) ?
+                (sz - offsetof(Object, entry_array.compact)) / sizeof(uint32_t) :
+                (sz - offsetof(Object, entry_array.items)) / sizeof(uint64_t);
```
should use `sizeof(entry_array.compact[0])` and `sizeof(entry_array.items[0])`

daandemeyer replied:
This would be a bigger change I think. There's nothing in the coding style about this and the rest of the file is using this style so I prefer to stick to that.

daandemeyer committed:
journal: Don't allocate objects above UINT32_MAX in compact mode

To allow storing offsets as 32-bit, we should never allocate objects
outside of the 32-bit range.

daandemeyer committed:
journal: Reduce space allocated for data, trie hash tables in compact mode

In compact mode, because we inline MESSAGE objects into the Entry object,
we end up fewer data objects per file, about 2/3 of the amount of entries
in a non-compact journal. Let's reduce the size of the data hash table
and trie hash table accordingly.

Number of data entries per file for system journal in compact mode:

/home/daandemeyer/projects/systemd/tmp/copy@d5f0b421e7264c3e9e4b10780029fcd0-0000000000000001-0005cdd5d27d2d94.journal: 113311
/home/daandemeyer/projects/systemd/tmp/copy@d5f0b421e7264c3e9e4b10780029fcd0-000000000004962b-0005ce4384695cd4.journal: 102223
/home/daandemeyer/projects/systemd/tmp/copy@d5f0b421e7264c3e9e4b10780029fcd0-0000000000298d1a-0005cfadbe650a0c.journal: 93671
/home/daandemeyer/projects/systemd/tmp/copy@d5f0b421e7264c3e9e4b10780029fcd0-00000000002053b6-0005cf6655ddb579.journal: 90780
/home/daandemeyer/projects/systemd/tmp/copy@d5f0b421e7264c3e9e4b10780029fcd0-00000000001bb51b-0005cf3a84e7a3bd.journal: 89281
/home/daandemeyer/projects/systemd/tmp/copy@d5f0b421e7264c3e9e4b10780029fcd0-00000000000927c6-0005ce9eb78f1b15.journal: 85873
/home/daandemeyer/projects/systemd/tmp/copy@d5f0b421e7264c3e9e4b10780029fcd0-00000000000dd673-0005cece4242847b.journal: 84350
/home/daandemeyer/projects/systemd/tmp/copy.journal: 79482
/home/daandemeyer/projects/systemd/tmp/copy@d5f0b421e7264c3e9e4b10780029fcd0-0000000000127578-0005cef1afc242ec.journal: 77679
/home/daandemeyer/projects/systemd/tmp/copy@d5f0b421e7264c3e9e4b10780029fcd0-000000000024f770-0005cf8941801623.journal: 75602
/home/daandemeyer/projects/systemd/tmp/copy@d5f0b421e7264c3e9e4b10780029fcd0-0000000000171c33-0005cf15e3b8d49c.journal: 71032

Number of data entries per file for system journal in non-compact mode:

/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-000000000001750b-0005ce128e01000b.journal: 155649
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-00000000001dddfb-0005cf3dd5768f32.journal: 154270
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-000000000024e2ea-0005cf78e9c5ae9a.journal: 154196
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-000000000012095c-0005cee439c98426.journal: 153614
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-000000000030e309-0005cfdf9cbf0ee7.journal: 153251
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-00000000002e9274-0005cfc7147577e8.journal: 152944
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-00000000000fa98f-0005ced350dafd2c.journal: 152311
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-0000000000062908-0005ce4caed15dcf.journal: 150689
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-0000000000203056-0005cf573dedbe4d.journal: 150480
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-0000000000333ab9-0005cfef9c957367.journal: 150430
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-0000000000146366-0005cef557e2d20d.journal: 150216
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-000000000022805d-0005cf683e77bc0a.journal: 149996
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-00000000002c2ebd-0005cfb62ab245c5.journal: 149568
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-000000000016c06e-0005cf0785762604.journal: 149369
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-0000000000274ab8-0005cf8a01f4b69a.journal: 148248
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-000000000029b8b2-0005cf9b05613234.journal: 147902
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-00000000000d564d-0005cec1af9c5432.journal: 147346
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-000000000003d58e-0005ce3afd8524e2.journal: 147255
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-00000000000af51a-0005ceb0d3b90d23.journal: 146798
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-0000000000191537-0005cf196a6c4a8a.journal: 146645
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-00000000001b844e-0005cf2bbc072bd5.journal: 146572
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-000000000009bc02-0005ce9c95ee4584.journal: 83220
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system.journal: 69055
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-000000000008cbdc-0005ce5ecffdd340.journal: 66396
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-0000000000009c46-0005ce0e3abbf007.journal: 56187
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-0000000000000001-0005cdd5d27d2d94.journal: 54650
/var/log/journal/9bc4fc1fd8c442b9817cd6fd77e03de6/system@2d0a2848c14f4df79fd48b9649fb2256-00000000002c2686-0005cfb628639e9f.journal: 5636

Before:

SYSTEMD_JOURNAL_NON_INDEXED_FIELDS="_CAP_EFFECTIVE _PID _UID _GID _SELINUX_CONTEXT _SOURCE_REALTIME_TIMESTAMP _MACHINE_ID _HOSTNAME _STREAM_ID _LINE_BREAK SYSLOG_FACILITY SYSLOG_IDENTIFIER _COMM _CMDLINE _EXE _SYSTEMD_INVOCATION_ID"

OBJECT TYPE	 ENTRIES SIZE
Unused           0	 0B
Data             985433  92.3M
Field            2142    115.7K
Entry            3500812 614.4M
Data Hash Table  9	 32.0M
Field Hash Table 9	 46.9K
Entry Array	 240841  191.2M
Tag              0	 0B
Trie Node        1752652 66.8M
Trie Hash Table  9	 47.9M
Boot ID          55	 1.7K
Total            6481962 1.0G

After:

SYSTEMD_JOURNAL_NON_INDEXED_FIELDS="_CAP_EFFECTIVE _PID _UID _GID _SELINUX_CONTEXT _SOURCE_REALTIME_TIMESTAMP _MACHINE_ID _HOSTNAME _STREAM_ID _LINE_BREAK SYSLOG_FACILITY SYSLOG_IDENTIFIER _COMM _CMDLINE _EXE _SYSTEMD_INVOCATION_ID"

OBJECT TYPE      ENTRIES SIZE
Unused           0       0B
Data             980599  92.0M
Field            1948    105.2K
Entry            3501263 614.5M
Data Hash Table  8       18.9M
Field Hash Table 8       41.7K
Entry Array      238474  188.0M
Tag              0       0B
Trie Node        1753996 66.9M
Trie Hash Table  8       28.4M
Boot ID          53      1.6K
Total            6476357 1009.0M

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1690,9 +1704,11 @@ static int link_entry_into_array(JournalFile *f,
-        r = journal_file_append_object(f, OBJECT_ENTRY_ARRAY,
-                                       offsetof(Object, entry_array.items) + n * sizeof(uint64_t),
-                                       &o, &q);
+        const uint64_t sz = JOURNAL_HEADER_COMPACT(f->header) ?
```
We usually do not use `const` for variables.

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1702,7 +1718,10 @@ static int link_entry_into_array(JournalFile *f,
+        if (JOURNAL_HEADER_COMPACT(f->header))
+                o->entry_array.compact[i] = htole32(p);
+        else
+                o->entry_array.items[i] = htole64(p);
```
Maybe, better to introduce a new helper for this?

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -3632,21 +3632,16 @@ int journal_file_copy_entry(JournalFile *from, JournalFile *to, Object *o, uint6
                         return r;
 
-                if (le_hash != o->data.hash)
-                        return -EBADMSG;
```
Why these are dropped? Could you explain in the commit message?

daandemeyer replied:
Added a detailed commit message

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1301,7 +1301,7 @@ int journal_file_find_data_object_with_hash(
                 const void *data, uint64_t size, uint64_t hash,
```
The commit message: `journal_file_entry_next()` -> `journal_file_entry_item_next()`

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1605,18 +1577,192 @@ static int journal_file_append_data(
+                return -E2BIG;
 
-        sz = le64toh(READ_NOW(o->object.size));
+        if (compression) {
```
`compression` is not a boolean:
```
if (compression != 0) {
```

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1605,18 +1577,192 @@ static int journal_file_append_data(
-                return 0;
+        /* We can't read objects larger than 4G on a 32bit machine */
+        t = (size_t) size;
+        if ((uint64_t) t != size)
```
It is not necessary to introduce `t`:
```
if ((uint64_t) (size_t) size != size)
       return -E2BIG;
```

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1336,40 +1336,12 @@ int journal_file_find_data_object_with_hash(
-                           memcmp(o->data.payload, data, size) == 0) {
+                r = journal_file_data_payload(f, o, p, NULL, 0, 0, &d, &rsize);
+                if (r < 0)
+                        return r;
```
`d` and `rsize` not set if `r == 0`.

Initialize them in the above before calling `journal_file_data_payload()`, or conditionalize the call of `memcmp_nn()` below with `r > 0`.

daandemeyer replied:
journal_file_data_payload() will never return 0 if field and field_length are NULL and 0 respectively but it doesn't hurt to be safe. Will add initialization for both variables.

yuwata replied:
Ah, indeed. Then, maybe it is better to add `assert(r > 0);`.

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1605,18 +1577,192 @@ static int journal_file_append_data(
+        assert(!e || e->object.type == OBJECT_ENTRY);
+        assert(offset);
+        assert(i);
+        assert(!!field == !!field_length);
```
Not a matter, but if we follow our coding style strictly, this should be the following as `field_length` is not a boolean:
```
assert(!field == (field_length == 0));
```

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -3628,51 +3774,29 @@ int journal_file_copy_entry(JournalFile *from, JournalFile *to, Object *o, uint6
+                        break;
+
+                n++;
+        }
```
Why is `journal_file_entry_n_items()` replaced with this loop??

daandemeyer replied:
With these changes, journal_file_entry_n_items() becomes O(n) instead of O(1) so it seemed like a good idea to remove it to discourage calculating the size of the entry items unless it's absolutely necessary since it's a more expensive operation now..

yuwata replied:
Ah, the function is dropped by this commit. I did not notice that. Thanks.

So, how heavy is this? What is the typical value of `n`?
Maybe it is better to drop the first loop and replace `items = newa(EntryItemEx, n);` with `GREEDY_REALLOC()`?

daandemeyer replied:
Typical value is between 15-30 I'd guess. I didn't notice any slowdowns during testing. Also, this function is only invoked by the FlushToVar journald varlink method, which I don't think is on the hot path.

yuwata commented on src/libsystemd/sd-journal/journal-verify.c:
```diff
@@ -275,13 +275,20 @@ static int journal_file_object_verify(JournalFile *f, uint64_t offset, Object *o
-                                      "Invalid entry item (%"PRIu64"/%"PRIu64" offset: "OFSfmt,
-                                      i, journal_file_entry_n_items(o),
-                                      le64toh(o->entry.items[i].object_offset));
+                for (i = 0;;) {
```
Maybe it is better to declare `i` for each loop?

daandemeyer replied:
Will do in a separate commit

yuwata commented on src/libsystemd/sd-journal/journal-def.h:
```diff
@@ -227,12 +248,17 @@ enum {
         /* Added in 246 */                              \
         le64_t data_hash_chain_depth;                   \
         le64_t field_hash_chain_depth;                  \
+        /* Added in vnext */                            \
```
Let's include this in v251!

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1605,18 +1577,190 @@ static int journal_file_append_data(
+                                        object_compressed_to_string(compression),
+                                        size);
+
+                        found = r > 0;
```
Maybe, it is better to return earlier if `r == 0`, and drop `found`:
```
if (field) {
        r = decompress_startswith();
        if (r < 0)
                return log_debug_errno(r);
        if (r == 0)
                return 0;
}

size_t rsize;
r = decompress_blob();
...
```

daandemeyer replied:
Yeah, this turned out much nicer. Thanks for the suggestion!

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1605,18 +1577,190 @@ static int journal_file_append_data(
+                        found = true;
+
+                if (found) {
+                        size_t rsize = 0;
```
The initialization is necessary?

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1605,18 +1577,190 @@ static int journal_file_append_data(
+                if (field)
+                        found = size >= field_length+1 &&
+                           memcmp(payload, field, field_length) == 0 &&
+                           payload[field_length] == '=';
```
Also here, I think the code become simpler by returning earlier if one of these conditions is false.

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -3282,6 +3283,14 @@ int journal_file_open(
+        r = getenv_bool("SYSTEMD_JOURNAL_COMPACT");
+        if (r < 0) {
+                if (r != -ENXIO)
+                        log_debug_errno(r, "Failed to parse $SYSTEMD_JOURNAL_COMPACT environment variable, ignoring.");
```
`ignoring.` -> `ignoring: %m`

yuwata reviewed (approved):
LGTM. But I am not sure if the last half commits are reasonable, and I only briefly looked them. I'd like to leaves to merge this PR for other reviewers.

daandemeyer committed:
journal: Run unit tests with and without compact mode enabled

medhefgo commented on src/libsystemd/sd-journal/journal-def.h:
```diff
@@ -118,7 +118,10 @@ struct HashTableObject {
-        le64_t items[];
+        union {
+                le64_t items[0];
+                le32_t compact[0];
```
Why not `le32_t compact[];`? That would be standard c code (and might not result in those false positives).

daandemeyer replied:
Lennart mentioned this as well but flexible arrays are not allowed in unions in standard C. And wrapping them in an anonymous struct doesn't work because you can't have a struct with only a flexible array member field.

daandemeyer committed:
journal: Turn adding entry array items into a O(1) operation from O(n)

Previously, we'd iterate an entry array from start to end every time
we added an entry offset to it. To speed up this operation, we cache
the last entry array in the chain and how many items it contains.
This allows the addition of an entry to the chain to be done in
constant time instead of linear time as we don't have to iterate
the entire chain anymore every time we add an entry.

This reduces the time to copy my laptop's 4G journal from 70s before
this commit to 43s after this commit.

daandemeyer committed:
journal: Add --convert= command to journalctl

--convert writes the journal files read by journalctl to the given
location. The location should be specified as a full journal file
path (e.g. /a/b/c/converted.journal). The directory specifies where
the converted journal files will be stored. The filename specifies
the naming convention the converted journal files will follow.

poettering commented on src/journal/journalctl.c:
```diff
@@ -2127,6 +2138,54 @@ static int wait_for_change(sd_journal *j, int poll_fd) {
+                return -ENOMEM;
+
+        r = managed_journal_file_open(-1, arg_convert, O_RDWR | O_CREAT, 0640, true, UINT64_MAX, false,
+                                      &(JournalMetrics) { -1, -1, -1, -1, -1, -1 }, mmap, NULL, NULL, &to);
```
hmm "-1" (better: UINT64_MAX) means automatic metrics. Here you don#t want any metrics, i.e. no rotation, hence set all of these metrics to zero? (which is equivalent to passing NULL)

daandemeyer replied:
I'm using this command for comparing journal sizes between regular and compact journals so having some rotation seemed like a better idea than having no rotation at all. 

This seems like something that's minor though, it'll still be useful without rotation, and I agree it makes more sense for the command itself to not do rotation, so will fix

poettering commented on src/journal/journalctl.c:
```diff
@@ -2127,6 +2138,54 @@ static int wait_for_change(sd_journal *j, int poll_fd) {
+                if (!journal_shall_try_append_again(to->file, r))
+                        return log_error_errno(r, "Can't write entry: %m");
+
+                log_info("Rotating journal.");
```
hmm, do we really want to do rotation here? sounds like a separate problem to me.

poettering commented on src/shared/journal-util.c:
```diff
@@ -137,3 +137,52 @@ int journal_access_check_and_warn(sd_journal *j, bool quiet, bool want_other_use
         return r;
 }
+
+bool journal_shall_try_append_again(JournalFile *f, int r) {
```
maybe mention that these are return codes from journal_file_open()

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -556,6 +556,10 @@ static int journal_file_allocate(JournalFile *f, uint64_t offset, uint64_t size)
                 return -E2BIG;
 
+        /* Refuse to go over 4G in compact mode so offsets can be stored in 32-bit. */
+        if (JOURNAL_HEADER_COMPACT(f->header) && offset + size > UINT32_MAX)
```
shouldn#t this be `new_size > UINT32_MAX`?

poettering commented on src/libsystemd/sd-journal/journal-def.h:
```diff
@@ -118,7 +118,10 @@ struct HashTableObject {
         le64_t next_entry_array_offset;
-        le64_t items[];
+        union {
+                le64_t items[0];
```
maybe rename this one too, in order to avoid usage mistakes

poettering replied:
e.g. `regular[]`?

poettering replied:
or large[]?

poettering commented on src/libsystemd/sd-journal/journal-def.h:
```diff
@@ -118,7 +118,10 @@ struct HashTableObject {
-        le64_t items[];
+        union {
+                le64_t items[0];
+                le32_t compact[0];
```
why do you size this to zero btw? why not just []?

daandemeyer replied:
Flexible array members in unions are not allowed. We have to use the GCC zero sized array extension to make it work.

poettering commented on src/libsystemd/sd-journal/journal-def.h:
```diff
@@ -118,7 +118,10 @@ struct HashTableObject {
+        union {
+                le64_t items[0];
+                le32_t compact[0];
+        };
```
i'd give this union the old name, i.e. `items`. that way you can still use it for offsetof()

daandemeyer replied:
This actually ends up simplifying things a lot

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -763,9 +763,9 @@ static int journal_file_check_object(JournalFile *f, uint64_t offset, Object *o)
-                    (sz - offsetof(Object, entry_array.items)) / sizeof(le64_t) <= 0)
+                if (sz < journal_file_entry_array_items_offset(f) ||
+                    (sz - journal_file_entry_array_items_offset(f)) % journal_file_entry_array_item_size(f) != 0 ||
+                    (sz - journal_file_entry_array_items_offset(f)) / journal_file_entry_array_item_size(f) <= 0)
```
I'd drop journal_file_entry_array_items_offset() and just name the union, see above.

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1684,6 +1692,16 @@ uint64_t journal_file_hash_table_n_items(Object *o) {
+        assert(o);
+
+        if (JOURNAL_HEADER_COMPACT(f->header))
+                o->entry_array.compact[i] = htole32(p);
```
please throw an error here if p > UINT32_MAX.

(or, maybe just add an assert, in case this is impossible anyway)

poettering commented on src/libsystemd/sd-journal/journal-file.h:
```diff
@@ -208,6 +209,14 @@ int journal_file_append_entry(
 
+static inline size_t journal_file_entry_array_items_offset(JournalFile *f) {
+        return JOURNAL_HEADER_COMPACT(f->header) ? offsetof(Object, entry_array.compact) : offsetof(Object, entry_array.items);
+}
```
as mentioned, I'd drop this helper. i.e. give the union a name, and you can use offsetof directly on it

poettering commented on src/libsystemd/sd-journal/journal-def.h:
```diff
@@ -91,6 +91,12 @@ struct EntryItem {
+typedef struct {
+        uint64_t object_offset;
+        uint64_t hash;
+} EntryItemEx;
```
hmm, this is purely a runtime thing? if so, don't put it in journal-def.h. This file should be about the file format and nothing else, and if this is not something that ends up in the file format it shouldn't be here.

poettering replied:
also, maybe add an explicit comment pointing to the fact that this is NE, not LE, unlike Entryitem!

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -3378,6 +3379,14 @@ int journal_file_open(
+                        log_debug_errno(r, "Failed to parse $SYSTEMD_JOURNAL_COMPACT environment variable, ignoring: %m");
+                f->compact = true;
+        } else
+                f->compact = r;
```
it think we shouldn't default to compat mode if the metrics' `max_size` is larger than UINT32_MAX. because people tell us that way they want larger journal files, and in that case we should use large offsets.

daandemeyer replied:
I guess for compat this is probably a good idea. But maybe we should also log a warning that they really should reduce their max size so they can benefit from compact mode.

daandemeyer replied:
Why don't we just clamp the max file size to 4G if compact mode is enabled (and maybe log that we're changing the user configured default value?)

poettering commented on src/libsystemd/sd-journal/sd-journal.c:
```diff
@@ -2477,7 +2343,21 @@ _public_ int sd_journal_enumerate_available_data(sd_journal *j, const void **dat
+
+                /* Try with the next field */
+                r = journal_file_entry_item_next(
+                        j->current_file,
```
our usual coding style says for continuation lines to use 16 space indenting

poettering replied:
(here and everywhere else)

poettering commented on src/libsystemd/sd-journal/journal-def.h:
```diff
@@ -95,17 +98,22 @@ struct EntryItem {
         uint64_t object_offset;
         uint64_t hash;
+        /* The hash used to calculate the Entry object's XOR hash field. */
+        uint64_t xor_hash;
```
the existing `hash` field should also get a comment if you now have two `hash` fields in the struct

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1796,6 +2031,60 @@ int journal_file_entry_item_next(
+        return JOURNAL_HEADER_COMPACT(f->header) ?
+                journal_file_entry_item_next_compact(
+                        f, e, offset, i, field, field_length, data_threshold, ret_offset, ret_data, ret_size) :
+                journal_file_entry_item_next_non_compact(
```
maybe calll "non_compact" just "regular"

poettering commented on src/libsystemd/sd-journal/journal-def.h:
```diff
@@ -47,11 +47,12 @@ typedef enum ObjectType {
+        OBJECT_COMPRESSED_ZSTD  = 1 << 2,
         OBJECT_COMPRESSION_MASK = (OBJECT_COMPRESSED_XZ | OBJECT_COMPRESSED_LZ4 | OBJECT_COMPRESSED_ZSTD),
-        _OBJECT_COMPRESSED_MAX = OBJECT_COMPRESSION_MASK,
+        FIELD_UNIQUE            = 1 << 3,
```
should always carry the same prefix as the other flags

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -646,6 +647,10 @@ static uint64_t minimum_header_size(Object *o) {
 
+        if (o->object.type == OBJECT_ENTRY)
+                return JOURNAL_HEADER_COMPACT(f->header) ? offsetof(Object, entry.payload) :
+                                                           offsetof(Object, entry.items);
```
as earlier, just give the union a name, then this ternary op goes away

daandemeyer replied:
In this case, giving the union a name doesn't work since `offsetof(Object, entry.payload) != offsetof(Object, entry.items)`.

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1968,21 +1952,151 @@ static int journal_file_entry_item_next_compact(
+
+        for (p = *i; p < offset + sz;) {
+                uint8_t *d;
+                uint8_t flags;
```
merge with previous line?

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -1968,21 +1952,151 @@ static int journal_file_entry_item_next_compact(
+        p = *i == 0 ? le64toh(READ_NOW(e->entry.trie_offset)) : *i;
+
+        /* All of an entry's trie and data nodes are located before the entry object in the journal file. */
+        if (p > offset + sz)
```
overflow check missing? (here and at many other places). afte rall we read the size from an untrusted source...

daandemeyer replied:
I fixed two missing overflow checks but I'm not sure if there's any more that I might have missed

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -2397,6 +2513,9 @@ static int journal_file_append_entry_internal(
         assert(ts);
 
+        if (!JOURNAL_HEADER_COMPACT(f->header))
+                assert(n_inlined == 0);
```
write as:

```
assert(JOURNAL_HEADER_COMPACT(f->header) || n_inlined == 0);
``

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -2428,18 +2547,49 @@ static int journal_file_append_entry_internal(
+                offsetof(Object, entry.payload) :
                 offsetof(Object, entry.items) + (n_items * sizeof(EntryItem));
 
+        for (unsigned i = 0; i < n_inlined; i++) {
```
size_t?

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -2428,18 +2547,49 @@ static int journal_file_append_entry_internal(
                 o->entry.trie_offset = htole64(parent_offset);
-        else
+
+                for (unsigned i = 0; i < n_inlined; i++) {
```
size_t

poettering commented on src/libsystemd/sd-journal/journal-def.h:
```diff
@@ -47,11 +47,12 @@ typedef enum ObjectType {
 
```
i don't grok this commit. if you drop "deduplication" as you call it, you also drop the *indexing* by this field. not indexing message is quite a limitiation i'd say...

i.e. i would assume that:

```
logger foo
journalctl MESSAGE=foo
```

always works. but with this change it really doesn't?

daandemeyer replied:
Is not indexing the message such a limitation though? I might be wrong, but doing exact matches on the MESSAGE field seems like a very rare use case. Instead, I'd expect most querying of MESSAGE fields to happen via the --grep option instead. Am I wrong here?

poettering replied:
i actually use this all the time. i.e. add a marker somehwere with a simple `log_notice("wuff")` then search for it with `journalctl _MESSAGE=wuff`.

daandemeyer replied:
Hmm, I don't mind keeping the default to index everything (which we probably should do regardless to keep compat intact) but it'd be nice to have the environment knob available to disable indexing for fields since all this indexing information does end up taking quite a bit of space.

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -326,7 +326,34 @@ static int journal_file_init_header(JournalFile *f, JournalFile *template) {
         return 0;
 }
 
+static int journal_file_refresh_boot_id(JournalFile *f, const sd_id128_t *boot_id) {
```
so, for this commit I'd really prefer a different approach: we already store the boot ID in the file header. This is currently the "last writer" boot ID. Let's introduce a new rule: when we open a file and notice it is from a different boot ID than our current, then take that as reason for rotation. And then, drop the boot ID from the EntryItem, and artificially synthesize _BOOT_ID= fields on all entries on read, and never store them at all, whatsoever. That way we'd then enforce the rule: journal files stem from a single boot only, and can never span multiple boots. Which I think is a good rule probably.

(Of course, this must be something that is indicated by a flag in the file, and that journald will enable but not the remoting stuff, as for that stuff multiple sources ight end up in the same journal file after all)

(And similar for the _MACHINE_ID= field, except that that one doesn't show up in EntryItem anyway, just as regular field.)

poettering replied:
and _HOSTNAME is similar too: but unlike the boot id/machine id this one can change during runtime, i.e. our writer must be slightly updated to rotate whenver the hostname is changed. We do something like that already when the clock jumps iirc, to guarantee the timestamp arrays remain perfectly sorted.

poettering replied:
(of course, adding such a logic would mean some special care needs to be taken when filtering by _BOOT_ID=/_MACHINE_ID=/_HOSTNAME=, since we have to skip whole files then, and never actually search for those fields.

poettering replied:
btw, along these lines, one more possible idea: it might work to reduce the size of the two usec timestamps in the entries from 64bit to 32bit. And then store a 64bit "base" value for each in the journal file header, so that the actual timestamps are always the sum of the base values plus the 32bit timestamps. And should the 32bit timestamps overrun, then use this as cause for rotation.

poettering replied:
(and similar for seqnum)

daandemeyer replied:
> (Of course, this must be something that is indicated by a flag in the file, and that journald will enable but not the remoting stuff, as for that stuff multiple sources ight end up in the same journal file after all)

Adding another flag means a substantial amount of extra complexity, since we'll end up with 4 different versions of the Entry object struct (regular x multiple boot ids, regular x single boot id, compact x multiple boot ids, compact x single boot id).  I'd really like to keep things behind a single flag if possible. What if we put this behind compact mode as well? That would mean we can't use compact mode in combination with the journal remote stuff, but is that really a problem?

poettering reviewed (changes requested):
sorry for taking so long for the review.

I like the first parts a lot more than the stuff that drops indexing/dedup from some fields. Could we please do that in two steps?

daandemeyer committed:
journal: Add support for not deduplicating specific fields in compact mode

For fields that are almost always unique, allocating a separate
Data object ends up adding a noticeable amount of overhead. By
adding support for storing these fields inline in the entry object,
we reduce the space required to store these unique fields.

If a field is marked as unique, we don't allocate a Data object and
store it inline in the Entry instead. The entry payload for storing
a single unique field has the following format:

- 1 byte for flags (compressed, etc)
- 4 bytes for size
- data (optionally compressed)

Each unique field is serialized to this format and the concatenation
of all the serialized unique fields becomes the entry payload.

When iterating over entries, we first iterate over all the deduplicated
fields via the trie. Once those are done, we continue with the inlined
fields. journal_file_entry_next()'s implementation is extended to support
this. One change in it's API is that ret_offset is set to zero for inline
fields if it is provided.

The list of fields to not deduplicate can be configured with the
$SYSTEMD_JOURNAL_UNIQUE_FIELDS environment variable. If it's not set, we
default to deduplicating all fields except MESSAGE.

With this change, we need to have the field object available before we
append the data object so we move field object allocation out of
journal_file_append_data() and into journal_file_append_entry() and
journal_file_copy_entry() instead.

Before:

OBJECT TYPE      ENTRIES  SIZE
Unused           0        0B
Data             3521895  587.0M
Field            3140     169.4K
Entry            3499118  240.2M
Data Hash Table  14       49.7M
Field Hash Table 14       73.0K
Entry Array	 577350   499.5M
Tag              0        0B
Trie Node        5767903  220.0M
Trie Hash Table  14       74.6M
Total            13369448 1.6G

After:

OBJECT TYPE      ENTRIES SIZE
Unused           0       0B
Data             1022925 95.3M
Field            2808    151.3K
Entry            3499976 667.7M
Data Hash Table  13      46.2M
Field Hash Table 13      67.8K
Entry Array      492907  576.7M
Tag              0       0B
Trie Node        1758648 67.0M
Trie Hash Table  13      69.3M
Total            6777303 1.4G

daandemeyer committed:
journal: Add support for not indexing specific fields in compact mode

Entry arrays remain one of the big contributors to journal
disk usage (even after introducing 32-bit offsets). To
improve the situation, let's add an environment variable
$SYSTEMD_JOURNAL_NON_INDEXED_FIELDS that allows configuring
which fields should not be indexed.

Data object of fields that are not indexed simply don't have
any entry offsets added to their entry arrays. This has the
effect that when used as matches, these data objects don't
return any entry matches at all.

Before:

OBJECT TYPE      ENTRIES SIZE
Unused           0       0B
Data             1022925 95.3M
Field            2808    151.3K
Entry            3499976 667.7M
Data Hash Table  13      46.2M
Field Hash Table 13      67.8K
Entry Array      492907  576.7M
Tag              0       0B
Trie Node        1758648 67.0M
Trie Hash Table  13      69.3M
Total            6777303 1.4G

After:

SYSTEMD_JOURNAL_NON_INDEXED_FIELDS="_CAP_EFFECTIVE _PID _UID _GID _SELINUX_CONTEXT _SOURCE_REALTIME_TIMESTAMP _MACHINE_ID _HOSTNAME _STREAM_ID _LINE_BREAK SYSLOG_FACILITY SYSLOG_IDENTIFIER _COMM _CMDLINE _EXE _SYSTEMD_INVOCATION_ID"

OBJECT TYPE      ENTRIES SIZE
Unused           0       0B
Data             987596  92.5M
Field            2127    114.8K
Entry            3500464 667.8M
Data Hash Table  9       32.0M
Field Hash Table 9       46.9K
Entry Array      241932  200.3M
Tag              0       0B
Trie Node        1752229 66.8M
Trie Hash Table  9       47.9M
Total            6484375 1.0G

daandemeyer committed:
journal: Split entry object information into compact/regular unions

Preparation for changing entry object timestamp information from
64-bit to 32-bit. We add accessor functions for all the entry object
fields so we can change how they're stored in one place without
having to modify all the use sites.

journal: Add compact/regular variant of entry seqnum field

Prep work for making the seqnum a 32-bit field in compact mode.

daandemeyer committed:
journal: Store entry seqnum as offset to the head entry

Saves 4 bytes per entry object.

daandemeyer committed:
journal: Drop boot ID storage in entry objects in compact mode

Instead, we limit journal files to a single boot ID in compact mode
which allows us to store the boot ID once in the journal header
instead of storing it per entry object.
